core/shared/platform/zephyr/zephyr_file.c

```diff
@@ -0,0 +1,975 @@
+ * descriptors.
+ * To pass the file descriptor reference to the higher level abstraction, we
+ * pass the index of the fd table to an `os_file_handle` struct.
+ * Then in the WASI implementation layer we can retrieve the file descriptor
+ * reference.
+ * We also fake the stdin, stdout and stderr file descriptors.
+ * We redirect the write operation on stdin, stdout and stderr to `os_printf`.
+ * We do not handle write on stdin and read on stdin, stdout and stderr.
+ */
+
+// No OS API wrapper (Zephyr):
+// file:
+//     off_t fs_tell(struct fs_file_t *zfp)
+// file system:
+//     int fs_mount(struct fs_mount_t *mp)
+//     int fs_unmount(struct fs_mount_t *mp
+//     int fs_readmount(int *index, const char **name)
+//     int fs_statvfs(const char *path, struct fs_statvfs *stat)
+//     int fs_mkfs(int fs_type, uintptr_t dev_id, void *cfg, int flags)
+//     int fs_register(int type, const struct fs_file_system_t *fs)
+//     int fs_unregister(int type, const struct fs_file_system_t *fs)
+
+// We will take the maximum number of open files
+// from the Zephyr POSIX configuration
+#define CONFIG_WASI_MAX_OPEN_FILES 16
+
+// Macro to retrieve a file system descriptor and check it's validity.
+#define GET_FILE_SYSTEM_DESCRIPTOR(fd, ptr)         \
+    do {                                            \
+        k_mutex_lock(&desc_array_mutex, K_FOREVER); \
+        ptr = &desc_array[(int)fd];                 \
+        if (!ptr) {                                 \
+            k_mutex_unlock(&desc_array_mutex);      \
+            return __WASI_EBADF;                    \
+        }                                           \
+        k_mutex_unlock(&desc_array_mutex);          \
+    } while (0)
+
+// Array to keep track of file system descriptors.
+static struct zephyr_fs_desc desc_array[CONFIG_WASI_MAX_OPEN_FILES];
+
+// mutex to protect the file descriptor array
+K_MUTEX_DEFINE(desc_array_mutex);
+
+static char prestat_dir[MAX_FILE_NAME + 1];
+
+static struct zephyr_fs_desc *
+zephyr_fs_alloc_obj(bool is_dir, const char *path, int *index)
+{
+    int i;
+    struct zephyr_fs_desc *ptr = NULL;
+    *index = -1; // give a default value to index in case table is full
+
+    k_mutex_lock(&desc_array_mutex, K_FOREVER);
+    for (i = 0; i < CONFIG_WASI_MAX_OPEN_FILES; i++) {
+        if (desc_array[i].used == false) {
+            ptr = &desc_array[i];
+            ptr->used = true;
+            ptr->is_dir = is_dir;
+            ptr->path = strdup(path);
+            if (ptr->path == NULL) {
+                ptr->used = false;
+                k_mutex_unlock(&desc_array_mutex);
+                return NULL;
+            }
+            *index = i;
+            break;
+        }
+    }
+
+    k_mutex_unlock(&desc_array_mutex);
+
+    return ptr;
+}
+
+static inline void
+zephyr_fs_free_obj(struct zephyr_fs_desc *ptr)
+{
+    free(ptr->path);
+    ptr->path = NULL;
+    ptr->used = false;
```

Comment:
Seems that ptr is always a pointer pointing to an element of desc_array? Should also add `k_mutex_lock(&desc_array_mutex) ... k_mutex_unlock(&desc_array_mutex)` for these lines?

Reply:
And another issue is that will all allocated elements of desc_array be freed when application exits (e.g. wasm_runtime_destroy() is called)? For example, wasm app opened some files and some elements of desc_array are used and their path fields are allocated, and wasm app doesn't close them before exit.

Reply:
True, I didn't think about this case.

If we create a function to cleanup the `desc_array`, then we will have to call it in  either `wasm_runtime_destroy` or `wasm_runtime_destroy_wasi` .

But where should we declare such a function that is platform specific ? 

The simple approach would be to put it in the `platform_api_extension.h` and have it locked by `#ifdef BH_PLATFORM_ZEPHYR`. 

I don't think that we want platform  specific code in the plateform API. We could create a macro like `BH_PLATFORM_NEED_WASI_CLEANUP` and then use it to lock with `#ifdef` an `os_cleanup_wasi` or something along those lines.

Reply:
BTW, could you add `k_mutex_lock(&desc_array_mutex) ... k_mutex_unlock(&desc_array_mutex)` for these lines since `ptr->used` is set to false without protection?

Reply:
Thanks for the feedbacks, I will look into this.